### PR TITLE
MET-8 feat: Generate Textarea component

### PR DESCRIPTION
**Pull request by supreethavadhani**
## Add a textarea generator

The commit consists of rendering all the textarea component in the form using a single line of html. See screenshots below.

### Ticket - [MET-8](https://oregonstate-innovationlab.atlassian.net/browse/MET-8)

| Before  | After |
| ----------- | ----------- |
| Previously a developer could render only textboxes and not textareas using the framework  | Now, a developer can render textboxes and textarea in a form with a single line of HTML code.     |

## What you need to check for

1. Textarea is rendered on the screen.

Screen Shot 2022-07-05 at 9 39 47 PM

2. Only one line of HTML code is used to render textboxes and textarea 

Screen Shot 2022-07-05 at 10 25 31 PM

3. Field Validations work

### CODE REVIEW

**Comment by raghu-bhandi** [comment]
I suggest that you separate the 'frame-work' from 'Supreeth-work'. I understand that the modules you are using may not be available on npm, but it may be a good idea to split this project into two. A library that includes everything that you are NOT writing as part of this project ( and this is made available as an npm package) and an app that uses this library. You may take a decision about this based on the amount of effort involved in doing this..

**Comment by supreethavadhani** [comment]
> I suggest that you separate the 'frame-work' from 'Supreeth-work'. I understand that the modules you are using may not be available on npm, but it may be a good idea to split this project into two. A library that includes everything that you are NOT writing as part of this project ( and this is made available as an npm package) and an app that uses this library. You may take a decision about this based on the amount of effort involved in doing this..

Yes. That makes a lot of sense. Will figure out how to package the supreeth-work into a npm package framework 😁

File: src/app/framework-modules/elements/ct-field-generator/component.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	templateUrl: './component.html',
+	styleUrls: []
+})
+
+/**
+ * Automated field Generator component
+ * @fd - Formdata
+ */
+export class CTFieldGenerator {
+	@Input() public fd: FormData;
```
Comment (wbraynen) [inline src/app/framework-modules/elements/ct-field-generator/component.ts]: 1. Re. "Automated field Generator component": I have no idea what that means 😅, but maybe someone using your framework would know because they'd have the context?  If not, could we put a better description in here?  (Also, 3 adjectives for 1 noun is hard to parse; I would at least reword this to reorder the words:)
2. What is "form data"?  User input from an html form?
3. `fd` is not very self documenting.  At least name it `formData`, unless there is a better name.
4. The name of the class starts with the prefix `CT`.  Is this to namespace it?  If so, I thought in Java packaging is the mechanism for namespacing.  What does `CT` stand for?

File: src/app/framework-modules/elements/ct-field-generator/component.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+	fields:any;
+	form:Form;
+	control:FormControl;
```
Comment (wbraynen) [inline src/app/framework-modules/elements/ct-field-generator/component.ts]: Still wondering if you have a linter enabled.  This formatting is inconsistent with the above:
```
fd: FormData;
```
In one place there is a space after the colon, in the other place there is no space after the colon.

**Review by gangamamulya — APPROVED** [review]
Hi Supreeth, everything looks good. It can be more readable if you can add some comments. Other than that I don't think there is any major change.

File: src/app/framework-modules/elements/ct-field-generator/component.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	
+	ngOnInit() {
```
Comment (wbraynen) [inline src/app/framework-modules/elements/ct-field-generator/component.ts]: Another mysterious abbreviation.  What is `ng`?  Next generation?:). And why "OnInit"?  Is that an angular thing?

Could this function use a document comment and/or a more self-documenting name?

File: src/app/framework-modules/elements/ct-textarea/component.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	selector: 'ct-textarea',
+	templateUrl: './component.html',
+	styleUrls: []
+})
+
+/*
+* Wrapper class for angular material.
+* Unpacks values from the model to render a textarea
+* @ouput - valueChange - value change emitter.
+*/
```
Comment (wbraynen) [inline src/app/framework-modules/elements/ct-textarea/component.ts]: Should this comment start with `/**`?  Or it's not for javadoc?

File: src/app/framework-modules/elements/ct-textarea/component.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	@Input() public field: Field;
+	@Input() public fd: FormData;
+	@Output() public valueChange = new EventEmitter<any>();
+
+	control:FormControl;
+	
+	ngOnInit() {
+		this.control = this.fd.formGroup.get(this.field.name) as FormControl;
+	}
+	valueChangeDetector($event){
```
Comment (wbraynen) [inline src/app/framework-modules/elements/ct-textarea/component.ts]: Another linting nitpick.  Above it's `blah() {` but here it's `blah(){`. Inconsistent.

Please turn on linter and make it part of your PR checks, so the PR checks would fail if the new code isn't linted.  That way we won't have to leave linting comments on PRs.

File: src/app/framework-modules/elements/elements.ts [hunk 6]
```diff
@@ -224,15 +220,19 @@ import { CtHorizontalBarChartComponent } from './ct-horizontal-barchart/componen
...
         CtLinearGaugeComponent,
         CtHorizontalBarChartComponent,
+        CTFieldGenerator,
+        CTTextareaComponent,
         /**
          * Miscellaneous Exports
          */
         ReactiveFormsModule,
         CommonModule,
         FormsModule,
```
Comment (wbraynen) [inline src/app/framework-modules/elements/elements.ts]: The javadoc of typescript can pick up on this?  This should really be a `/**` and not just `//` or `/*`?

**Review by wbraynen — CHANGES_REQUESTED** [review]
Left some boring comments  😅

Seems you are using a javascript linter to lint typescript files?  Does eslint support typescript?  Also, if locally, maybe just lint staged files instead of the entire codebase?

Not super clear what this code does from reviewing it like this.  Could be more self-documenting and/or have better comments.

**Comment by wbraynen** [comment]
Also, nothing to QA for us?  I didn't find any QA instructions.

File: src/app/framework-modules/elements/ct-field-generator/component.ts (hunk 3 above)
Comment (supreethavadhani) [reply src/app/framework-modules/elements/ct-field-generator/component.ts]: ngOnint is an in-built angular lifecycle hook - https://angular.io/api/core/OnInit

**Comment by supreethavadhani** [comment]
The PR had a ton of file changes.
Reviewed code with @wbraynen in person.
Forced the merge because of the the 180,000 lines change.
